src/Microsoft.SqlTools.ServiceLayer/QueryExecution/DataStorage/SaveAsExcelFileStreamWriterHelper.cs

```diff
@@ -0,0 +1,583 @@
+            {
+                _referenceManager.IncreaseColumnReference();
+            }
+            /// <summary>
+            /// Write a string cell
+            /// </summary>
+            /// <param name="value"></param>
+            public void AddCell(string value)
+            {
+                _referenceManager.AssureColumnReference();
+                if (value == null)
+                {
+                    AddCellEmpty();
+                    return;
+                }
+
+                _writer.WriteStartElement("c");
+
+                _referenceManager.WriteAndIncreaseColumnReference();
+
+                _writer.WriteAttributeString("t", "inlineStr");
+
+                _writer.WriteStartElement("is");
+                _writer.WriteStartElement("t");
+                _writer.WriteValue(value);
+                _writer.WriteEndElement();
+                _writer.WriteEndElement();
+
+                _writer.WriteEndElement();
+            }
+            /// <summary>
+            /// Write a DateTime cell.
+            /// </summary>
+            /// <param name="dateTime">Datetime</param>
+            /// <remark>
+            /// If the DateTime do not have date part, it will be written as datetime and show as time only
+            /// If the DateTime is before 1900-03-01, save as string because excel doesn't support them.
+            /// Otherwise, save as datetime, and if the time is 00:00:00, show as yyyy-MM-dd.
+            /// Show the datetime as yyyy-MM-dd HH:mm:ss if none of the previous situations
+            /// </remark>
+            public void AddCell(DateTime dateTime)
+            {
+                _referenceManager.AssureColumnReference();
+                long ticks = dateTime.Ticks;
+                Style style = Style.DateTime;
+                double excelDate;
+                if (ticks < TicksPerDay) //date empty, time only
+                {
+                    style = Style.Time;
+                    excelDate = ((double)ticks) / (double)TicksPerDay;
+                }
+                else if (ticks < ExcelDateCutoffTick) //before excel cut-off, use string
+                {
+                    AddCell(dateTime.ToString("yyyy-MM-dd", System.Globalization.CultureInfo.InvariantCulture));
+                    return;
+                }
+                else
+                {
+                    if (ticks % TicksPerDay == 0) //time empty, date only
+                    {
+                        style = Style.Date;
+                    }
+                    excelDate = ((double)(ticks - ExcelEpochTick)) / (double)TicksPerDay;
+                }
+                AddCellDateTimeInternal(excelDate, style);
+            }
+
+            /// <summary>
+            /// Write a object cell
+            /// </summary>
+            /// The program will try to output number/datetime, otherwise, call the ToString 
+            /// <param name="o"></param>
+            public void AddCell(object o)
```

Comment:
It seems this code is duplicated in the SaveAsExcelFileStreamWriter. Can we eliminate on or the other?

Looking at the code I like the implementation of it here better. Maybe we can tweak this to accept a DbCellValue?